modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/AsyncSqlScriptCursorTest.java

```diff
@@ -0,0 +1,146 @@
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import org.apache.ignite.internal.sql.api.ResultSetMetadataImpl;
+import org.apache.ignite.internal.sql.engine.framework.NoOpTransaction;
+import org.apache.ignite.internal.testframework.IgniteTestUtils.RunnableX;
+import org.apache.ignite.internal.util.AsyncCursor;
+import org.apache.ignite.internal.util.AsyncCursor.BatchedResult;
+import org.apache.ignite.internal.util.AsyncWrapper;
+import org.apache.ignite.sql.ResultSetMetadata;
+import org.junit.jupiter.api.Test;
+
+/**
+ * Tests for {@link AsyncSqlScriptCursor}.
+ */
+public class AsyncSqlScriptCursorTest {
+    private static final int PAGE_SIZE = 1;
+
+    @Test
+    public void listenerIsCalledOnFirstPageOnly() throws Exception {
+        TestPageLoadListener listener = new TestPageLoadListener(() -> {});
+        AsyncCursor<Integer> cursor = prepareTestCursor(listener);
+
+        assertThat(listener.callsCounter.get(), is(0));
+
+        assertThat(cursor.requestNextAsync(PAGE_SIZE), willCompleteSuccessfully());
+        assertThat(listener.callsCountAsync(), willBe(1));
+
+        assertThat(fetchAll(cursor), is(true));
+        assertThat(listener.callsCountAsync(), willBe(1));
+    }
+
+    @Test
+    public void testErrorInPrefetchListenerDoesNotAffectExecution() {
+        TestPageLoadListener listener = new TestPageLoadListener(() -> {
+            throw new RuntimeException("Expected");
+        });
+
+        AsyncCursor<Integer> cursor = prepareTestCursor(listener);
+
+        // Ensure request future completes without errors.
+        assertThat(cursor.requestNextAsync(PAGE_SIZE), willCompleteSuccessfully());
+    }
+
+    @Test
+    public void testPrefetchListenerExecutesAsynchronously() throws Exception {
+        CountDownLatch latch = new CountDownLatch(1);
+        TestPageLoadListener listener = new TestPageLoadListener(() -> latch.await(30, TimeUnit.SECONDS));
+
+        // Make sure all data is fetched successfully.
+        assertThat(fetchAll(prepareTestCursor(listener)), is(true));
+
+        assertThat(listener.callsCountAsync(), willBe(0));
+
+        latch.countDown();
+
+        assertThat(listener.callsCountAsync(), willBe(1));
+    }
+
+    private static AsyncCursor<Integer> prepareTestCursor(PageLoadListener listener) {
+        AsyncCursor<Integer> data = new AsyncWrapper<>(IntStream.range(0, 10).boxed().collect(Collectors.toList()).iterator());
+        QueryTransactionWrapper tx = new QueryTransactionWrapper(NoOpTransaction.readOnly("TX"), true);
+        ResultSetMetadata rsMeta = new ResultSetMetadataImpl(Collections.emptyList());
+
+        return new AsyncSqlScriptCursor<>(SqlQueryType.QUERY, rsMeta, tx, data, listener);
+    }
+
+    private static boolean fetchAll(AsyncCursor<Integer> cursor) throws ExecutionException, InterruptedException, TimeoutException {
```

Comment:
I think we do not need the result of this method + this method can be rewritten into `do { .. } while(res.hasMore())`.
It also does not need timed variant of Future::get.

Reply:
Done

Reply:
Sorry, not done yet)

Reply:
Done